osquery/events/linux/auditdnetlink.cpp

```diff
@@ -51,166 +49,89 @@ bool ShouldHandle(const audit_reply& reply) noexcept {
-FLAG(bool,
-     disable_audit,
-     true,
-     "Disable receiving events from the audit subsystem");
-
 /// Control the audit subsystem by electing to be the single process sink.
 FLAG(bool, audit_persist, true, "Attempt to retain control of audit");
 
+/// Audit debugger helper
+HIDDEN_FLAG(bool, audit_debug, false, "Debug Linux audit messages");
+
 /// Control the audit subsystem by allowing subscriptions to apply rules.
 FLAG(bool,
      audit_allow_config,
      false,
      "Allow the audit publisher to change auditing configuration");
 
+/// Always uninstall all the audit rules that osquery uses when exiting
 FLAG(bool,
      audit_force_unconfigure,
      false,
      "Always uninstall all rules, regardless of whether they were already "
      "installed or not");
 
-/// Audit debugger helper
-HIDDEN_FLAG(bool, audit_debug, false, "Debug Linux audit messages");
+/// Forces osquery to remove all rules upon startup
+FLAG(bool,
+     audit_force_reconfigure,
+     false,
+     "Configure the audit subsystem from scratch");
 
-// External flags
+// External flags; they are used to determine which rules needs to be installed
+DECLARE_bool(audit_allow_fim_events);
 DECLARE_bool(audit_allow_process_events);
 DECLARE_bool(audit_allow_sockets);
-DECLARE_bool(audit_allow_fim_events);
 DECLARE_bool(audit_allow_user_events);
 
-bool IsAuditdNetlinkEnabled() noexcept {
-  if (FLAGS_disable_audit) {
-    return false;
-  }
-
-  return (FLAGS_audit_allow_process_events || FLAGS_audit_allow_sockets ||
-          FLAGS_audit_allow_fim_events || FLAGS_audit_allow_user_events);
-}
-
 enum AuditStatus {
   AUDIT_DISABLED = 0,
   AUDIT_ENABLED = 1,
   AUDIT_IMMUTABLE = 2,
 };
 
-AuditdNetlink& AuditdNetlink::get() {
-  static AuditdNetlink instance;
-  return instance;
-}
-
-bool AuditdNetlink::start() noexcept {
-  std::lock_guard<std::mutex> lock(initialization_mutex_);
-
-  if (initialized_)
-    return true;
+AuditdNetlink::AuditdNetlink() {
+  // Allocate the read buffer before we start the threads
+  const size_t read_buffer_size = 4096;
 
-  if (!IsAuditdNetlinkEnabled()) {
-    return true;
+  read_buffer_.resize(read_buffer_size);
+  if (read_buffer_.size() != read_buffer_size) {
+    VLOG(1) << "Failed to allocate the audit netlink read buffer";
+    throw std::bad_alloc();
   }
 
-  try {
-    // Allocate the read buffer before we start the threads
-    const std::size_t read_buffer_size = 4096;
-
-    read_buffer_.resize(read_buffer_size);
-    if (read_buffer_.size() != read_buffer_size) {
-      VLOG(1) << "Failed to allocate the audit netlink read buffer";
-      return false;
-    }
-
-    // Start the reading thread
-    std::packaged_task<bool(AuditdNetlink&)> recv_thread_task(
-        std::bind(&AuditdNetlink::recvThread, this));
-
-    recv_thread_.reset(
-        new std::thread(std::move(recv_thread_task), std::ref(*this)));
+  // Start the reading thread
+  std::packaged_task<bool(AuditdNetlink&)> recv_thread_task(
```

Comment:
I'm still concerned about this being its own task/thread. The preference is to use the `Dispatcher` for threads intended to run the duration of the process.

Reply:
I've already started working on this the other day! :+1: It's easier to do now since we unified the publishers

Reply:
Done! I pushed the change!